### ISSUE TITLE
for..in loop on a generator returns an empty string

### ISSUE THREAD

**Issue by bbqsrc** [mozilla/nunjucks#861]
When attempting to use a generator in place of an array, an empty string is returned.

``` javascript
function* generator() {
  yield 1; 
  yield 2; 
  yield 3;
}

nunjucks.renderString("{% for x in y %}{{ x }}{% endfor %}", { y: generator() })
// > ''
```

I expected it to behave like the following:

``` javascript
nunjucks.renderString("{% for x in y %}{{ x }}{% endfor %}", { y: [1, 2, 3] })
// > '123'
```

**Comment by vinnyrose** [comment]
This would be great for deferring certain operations until they are needed. And if they are not needed, then they won't be performed.

**Comment by noahlange** [comment]
So here's what Nunjucks kicks out as the compiled function for your example.

```javascript
function root(env, context, frame, runtime, cb) {
  var lineno = null;
  var colno = null;
  var output = "";
  try {
    var parentTemplate = null;
    frame = frame.push();
    var t_3 = runtime.contextOrFrameLookup(context, frame, "y");
    if (t_3) {
      var t_2 = t_3.length;
      for (var t_1 = 0; t_1 < t_3.length; t_1++) {
        var t_4 = t_3[t_1];
        frame.set("x", t_4);
        frame.set("loop.index", t_1 + 1);
        frame.set("loop.index0", t_1);
        frame.set("loop.revindex", t_2 - t_1);
        frame.set("loop.revindex0", t_2 - t_1 - 1);
        frame.set("loop.first", t_1 === 0);
        frame.set("loop.last", t_1 === t_2 - 1);
        frame.set("loop.length", t_2);
        output += runtime.suppressValue(t_4, env.opts.autoescape);
      }
    }
    frame = frame.pop();
    if (parentTemplate) {
      parentTemplate.rootRenderFunc(env, context, frame, runtime, cb);
    } else {
      cb(null, output);
    }
  } catch (e) {
    cb(runtime.handleError(e, lineno, colno));
  }
}

```
Looking at the code, we can see why generators won't work: generators don't have a length. Now, that isn't to suggest that Nunjucks couldn't switch to `for-of` in our compiled code, but the compiler would be very limited in the loop information it could output for a generator function.

```javascript
let idx = 0;
const iterable = y;
for (const item of iterable) {
  frame.set("x", item);
  frame.set("loop.index", idx + 1);
  frame.set("loop.index0", idx);
  frame.set("loop.first", idx === 0);
  // reverse index: ???
  // reverse 0-index: ???
  // last: ???
  // length: ???
  output += runtime.suppressValue(item, env.opts.autoescape);
  idx++;
}
```

Moreover, switching to `for-of` would mean losing compat through IE11 unless Nunjucks started transpiling outputted code for older browsers through Babel or whatever.

TL;DR: I'd love to see generator support happen eventually, but it'd require a pretty serious overhaul of loop compilation code, along with potentially-necessary transpilation of browser templates and loss of loop variables for some iterables.

**Comment by fdintino** [comment]
This will be fixed when #1053 is merged.

**Comment by fdintino** [comment]
Fixed with merge of #1053, in release v3.1.0
